[rebranch][Runtime] Remove attribute ordering hack

bnbarham opened the pull request:
Remove the split macro definitions and move the SWIFT_RUNTIME_EXPORT
macro to the beginning of the attributes list so that `extern "C"`
becomes the first attribute. This is possible since stable/20221013
where mixing GNU and standard attributes was fixed
(https://reviews.llvm.org/D137979).

Also remove an unnecessary include, which would prevent the compiler
itself being built on the released clang-15 (since the GNU/standard
attribute mixing fix isn't in clang-15).

bnbarham commented:
@swift-ci please test

compnerd reviewed (approved):
Beautiful!  Thank you!

buttaface commented:
Still failing on CI: I noticed that Saleem's patch was backported to stable/20221013, apple/llvm-project#5655, while [trunk still uses stable/20220421](https://github.com/apple/swift/blob/main/utils/update_checkout/update-checkout-config.json#L92), which would partially explain why this is failing (there appear to be other errors too).

bnbarham commented:
Interesting 🤔. I didn't think we needed Saleem's patch in stable/20220421 but ... apparently attributes are also broken there? It has quite a few conflicts so I'm not sure about cherry-picking it. Any thoughts @compnerd? Maybe we need to use this hack for `main` but switch `rebranch` to use this PR?

compnerd commented:
I think that leaving it on main for a little bit is fine.  We can fix this on the rebranch.  I can take a look at backporting the clang change to the current branch.  Seems better to fix it before the next release.

bnbarham committed:
[Runtime] Remove attribute ordering hack

Remove the split macro definitions and move the SWIFT_RUNTIME_EXPORT
macro to the beginning of the attributes list so that `extern "C"`
becomes the first attribute. This is possible since stable/20221013
where mixing GNU and standard attributes was fixed
(https://reviews.llvm.org/D137979).

Also remove an unnecessary include, which would prevent the compiler
itself being built on the released clang-15 (since the GNU/standard
attribute mixing fix isn't in clang-15).

bnbarham commented:
@swift-ci please test

compnerd commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -30,7 +30,6 @@
 #include "swift/ClangImporter/ClangModule.h"
 #include "swift/Demangling/ManglingMacros.h"
 #include "swift/IRGen/Linking.h"
-#include "swift/Runtime/HeapObject.h"
```
Was this meant to be rebranch only?  Seems like this can go on main as well?

bnbarham replied:
I can put it on `main` if you want, I was just going to leave it until the rebranch is merged.

bnbarham commented:
Note: we're apparently still failing on
```
extern "C" __attribute__((__visibility__("default"))) [[nodiscard]]
int myTest();
```

ie. mixed attributes with `extern "C"`.

bnbarham commented:
@swift-ci please test

bnbarham commented:
Re-running now that https://github.com/apple/llvm-project/pull/5869 is merged

bnbarham commented:
@swift-ci please test macOS platform

bnbarham commented:
@swift-ci please clean test

bnbarham committed:
Temporary debug to check why host compiler is being used

bnbarham commented:
@swift-ci please test Linux platform

bnbarham commented:
So that I don't keep forgetting the problem here:
  1. Clang 15 introduced a change that broke our attribute order. To workaround that we split `SWIFT_RUNTIME_EXPORT` into two.
  2. compnerd fixed that issue in newer Clang's, but we can't revert to the old use because the runtime headers are included in compiler libraries that are built with the host compiler (which could be Clang 15). We also can't change to moving `SWIFT_RUNTIME_EXPORT` to the start because that has always been broken in Clang (until comp nerd's fix).

The proper fix here will be to properly split the runtime headers - the function declarations should only be available to the runtime, which is required to be built with the just-built compiler. This would also help in not accidentally including compiler headers in the runtime.

Once that's done we can use `SWIFT_RUNTIME_EXPORT` again as the bootstrapped compiler will be the only compiler that touches those decls.
